File: frontend/src/api/llm.ts
import { ApiError, api } from './client'

export type LLMProvider = 'openai_compat' | 'anthropic' | 'ollama' | 'vllm' | 'llamacpp'
export type LLMIntent = 'salt_state' | 'ansible_playbook' | 'fleet_command' | 'explain' | 'fleet_query' | 'auto'

export interface LLMEndpoint {
  id: string
  name: string
  provider: LLMProvider
  base_url: string | null
  model: string
  max_tokens: number
  is_default: boolean
  enabled: boolean
  has_api_key: boolean
  created_at: string
  updated_at: string
  model_context_length: number | null
  model_capabilities: string[]
}

export interface LLMEndpointCreate {
  name: string
  provider: LLMProvider
  base_url?: string | null
  model: string
  max_tokens?: number
  is_default?: boolean
  enabled?: boolean
  api_key?: string | null
}

export type LLMEndpointUpdate = Partial<LLMEndpointCreate>

export interface LLMEndpointTestResult {
  ok: boolean
  latency_ms: number | null
  error: string | null
}

export interface ChatHistoryMessage {
  role: 'user' | 'assistant'
  content: string
}

export interface LLMQueryRequest {
  prompt: string
  intent: LLMIntent
  endpoint_id?: string | null
  history?: ChatHistoryMessage[]
}

export interface LLMQueryResponse {
  query_id: string
  intent: string
  result: string
  model_used: string
  endpoint_name: string
  input_tokens: number
  output_tokens: number
  duration_ms: number
}

export interface LLMQueryLogEntry {
  id: string
  intent: string
  prompt: string
  model_used: string | null
  duration_ms: number | null
  error: string | null
  created_at: string
}

export const llmApi = {
  list: () => api.get<LLMEndpoint[]>('/api/v1/llm/endpoints'),
  create: (data: LLMEndpointCreate) => api.post<LLMEndpoint>('/api/v1/llm/endpoints', data),
  update: (id: string, data: LLMEndpointUpdate) =>
    api.put<LLMEndpoint>(`/api/v1/llm/endpoints/${id}`, data),
  delete: (id: string) => api.delete(`/api/v1/llm/endpoints/${id}`),
  test: (id: string) =>
    api.post<LLMEndpointTestResult>(`/api/v1/llm/endpoints/${id}/test`),
  submitQuery: (data: LLMQueryRequest) => api.post<LLMQueryResponse>('/api/v1/llm/query', data),
  listQueries: () => api.get<LLMQueryLogEntry[]>('/api/v1/llm/queries'),
  discoverModels: (url: string, provider: string) =>
    api.post<{ models: Array<{ id: string; name: string; healthy: boolean; latency_ms: number | null }> }>(
      '/api/v1/llm/discover-models',
      { url, provider }
    ),
}

// ── Streaming query (SSE) ────────────────────────────────────────────────────
// EventSource cannot POST and cannot send custom headers (Authorization),
// so the streaming endpoint is consumed via fetch() + a ReadableStream
// reader. The wire format is SSE-compatible (each event is a single
// `data: <json>\n\n` line), which keeps it interoperable with curl, nginx
// proxy buffering, and any future generic SSE client.

export interface StreamDelta {
  type: 'delta'
  text: string
}

export interface StreamDone {
  type: 'done'
  query_id: string
  intent: string
  model_used: string
  endpoint_name: string
  input_tokens: number
  output_tokens: number
  duration_ms: number
  error?: string
}

export interface StreamError {
  type: 'error'
  error: string
}

export type StreamEvent = StreamDelta | StreamDone | StreamError

export interface StreamCallbacks {
  onDelta?: (text: string, fullText: string) => void
  onDone?: (final: StreamDone) => void
  onError?: (message: string) => void
}

/**
 * Stream an LLM query response via SSE. Returns an AbortController whose
 * .abort() method cancels both the client read loop and the upstream LLM
 * call (FastAPI propagates client disconnects).
 */
export function streamQuery(
  data: LLMQueryRequest,
  callbacks: StreamCallbacks,
): AbortController {
  const controller = new AbortController()
  void runStream(data, callbacks, controller).catch((err) => {
    if (controller.signal.aborted) return
    callbacks.onError?.(err instanceof Error ? err.message : String(err))
  })
  return controller
}

async function runStream(
  data: LLMQueryRequest,
  callbacks: StreamCallbacks,
  controller: AbortController,
): Promise<void> {
  const token = localStorage.getItem('access_token') || ''
  const res = await fetch('/api/v1/llm/query/stream', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'text/event-stream',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(data),
    signal: controller.signal,
  })

  if (!res.ok) {
    const body = await res.json().catch(() => ({}))
    throw new ApiError(
      res.status,
      (body as { detail?: string }).detail ?? res.statusText,
      (body as { error_code?: string }).error_code ?? null,
    )
  }
  if (!res.body) {
    throw new Error('Stream response had no body')
  }

  const reader = res.body.getReader()
  const decoder = new TextDecoder('utf-8')
  let buffer = ''
  let fullText = ''

  while (true) {
    const { value, done } = await reader.read()
    if (done) break
    buffer += decoder.decode(value, { stream: true })

    // SSE frames are separated by `\n\n`. Multiple frames may arrive in a
    // single chunk; partial trailing frames stay in the buffer for the
    // next iteration.
    let sep = buffer.indexOf('\n\n')
    while (sep !== -1) {
      const frame = buffer.slice(0, sep)
      buffer = buffer.slice(sep + 2)
      const line = frame.split('\n').find((l) => l.startsWith('data:')) || ''
      const payload = line.replace(/^data:\s*/, '')
      if (payload === '[DONE]') return
      if (!payload) {
        sep = buffer.indexOf('\n\n')
        continue
      }
      try {
        const ev = JSON.parse(payload) as StreamEvent
        if (ev.type === 'delta') {
          fullText += ev.text
          callbacks.onDelta?.(ev.text, fullText)
        } else if (ev.type === 'done') {
          callbacks.onDone?.(ev)
        } else if (ev.type === 'error') {
          callbacks.onError?.(ev.error)
        }
      } catch {
        // Ignore malformed frames; the server should never emit them but
        // a partial unicode boundary mid-frame is theoretically possible.
      }
      sep = buffer.indexOf('\n\n')
    }
  }
}

File: fleet_platform/schemas/llm.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

VALID_PROVIDERS = Literal["openai_compat", "anthropic", "ollama", "vllm", "llamacpp"]
VALID_INTENTS = Literal["salt_state", "ansible_playbook", "fleet_command", "explain", "fleet_query", "auto"]

# How the agent loop exposes tools to an endpoint (#710):
#   native    — provider-native tool-calling (OpenAI tools=[...])
#   anthropic — Anthropic tool_use blocks
#   json      — tool schema embedded in the prompt, JSON reply parsed out
#   none      — no tools exposed (plain chat / read-only Q&A)
VALID_TOOL_MODES = ("native", "json", "anthropic", "none")
ToolMode = Literal["native", "json", "anthropic", "none"]


def normalize_tool_mode(value: str | None) -> str:
    """Resolve a stored/legacy tool_mode to a safe value for the call path.

    Unknown or NULL values fall back to ``json`` (prompt-embedded), which works
    on every backend, so a legacy DB row never breaks a call.
    """
    return value if value in VALID_TOOL_MODES else "json"


class DiscoveredModel(BaseModel):
    id: str
    name: str
    healthy: bool
    latency_ms: int | None = None


class LLMEndpointCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    provider: VALID_PROVIDERS
    base_url: str = Field(..., min_length=1, max_length=500)
    api_key: str | None = None
    model: str = Field(..., min_length=1, max_length=255)
    max_tokens: int = Field(default=4096, ge=256, le=200000)
    is_default: bool = False
    enabled: bool = True
    model_context_length: int | None = None
    model_capabilities: list[str] | None = None
    tool_mode: ToolMode = "json"


class LLMEndpointUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    provider: VALID_PROVIDERS | None = None
    base_url: str | None = Field(default=None, min_length=1, max_length=500)
    api_key: str | None = None
    model: str | None = Field(default=None, min_length=1, max_length=255)
    max_tokens: int | None = Field(default=None, ge=256, le=200000)
    is_default: bool | None = None
    enabled: bool | None = None
    model_context_length: int | None = None
    model_capabilities: list[str] | None = None
    tool_mode: ToolMode | None = None


class LLMEndpointResponse(BaseModel):
    id: uuid.UUID
    name: str
    provider: str
    base_url: str
    has_api_key: bool
    model: str
    max_tokens: int
    is_default: bool
    enabled: bool
    created_at: datetime
    updated_at: datetime
    model_context_length: int | None = None
    model_capabilities: list[str] | None = None
    tool_mode: str = "json"

    model_config = {"from_attributes": True}


class LLMEndpointTestResponse(BaseModel):
    ok: bool
    latency_ms: int
    error: str | None = None


class ChatHistoryMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str = Field(..., max_length=4000)


class LLMQueryRequest(BaseModel):
    prompt: str = Field(..., min_length=1, max_length=8000)
    intent: VALID_INTENTS
    endpoint_id: uuid.UUID | None = None
    history: list[ChatHistoryMessage] = Field(default_factory=list, max_length=50)


class LLMQueryResponse(BaseModel):
    query_id: uuid.UUID
    intent: str
    result: str
    model_used: str
    endpoint_name: str
    input_tokens: int
    output_tokens: int
    duration_ms: int


class LLMQueryLogEntry(BaseModel):
    id: uuid.UUID
    intent: str
    prompt: str
    model_used: str | None
    duration_ms: int | None
    error: str | None
    created_at: datetime

    model_config = {"from_attributes": True}
